src/doc/trpl/guessing-game.md

```diff
@@ -0,0 +1,991 @@
+error: aborting due to previous error
+Could not compile `guessing_game`.
+```
+
+Whew! This is a big error. The core of it is that we have ‘mismatched types’.
+Rust has a strong, static type system. However, it also has type inference.
+When we wrote `let guess = String::new()`, Rust was able to infer that `guess`
+should be a `String`, and so it doesn’t make us write out the type. And with
+our `secret_number`, there are a number of types which can have a value
+between one and a hundred: `i32`, a thirty-two-bit number, or `u32`, an
+unsigned thirty-two-bit number, or `i64`, a sixty-four-bit number. Or others.
+So far, that hasn’t mattered, and so Rust defaults to an `i32`. However, here,
+Rust doesn’t know how to compare the `guess` and the `secret_number`. They
+need to be the same type. Ultimately, we want to convert the `String` we
+read as input into a real number type, for comparison. We can do that
+with three more lines. Here’s our new program:
+
+```rust,ignore
+extern crate rand;
+
+use std::io;
+use std::cmp::Ordering;
+use rand::Rng;
+
+fn main() {
+    println!("Guess the number!");
+
+    let secret_number = rand::thread_rng().gen_range(1, 101);
+
+    println!("The secret number is: {}", secret_number);
+
+    println!("Please input your guess.");
+
+    let mut guess = String::new();
+
+    io::stdin().read_line(&mut guess)
+        .ok()
+        .expect("failed to read line");
+
+    let guess: u32 = guess.trim().parse()
+        .ok()
+        .expect("Please type a number!");
+
+    println!("You guessed: {}", guess);
+
+    match guess.cmp(&secret_number) {
+        Ordering::Less    => println!("Too small!"),
+        Ordering::Greater => println!("Too big!"),
+        Ordering::Equal   => println!("You win!"),
+    }
+}
+```
+
+The new three lines:
+
+```rust,ignore
+    let guess: u32 = guess.trim().parse()
+        .ok()
+        .expect("Please type a number!");
+```
+
+Wait a minute, I thought we already had a `guess`? We do, but Rust allows us
+to ‘shadow’ the previous `guess` with a new one. This is often used in this
+exact situation, where `guess` starts as a `String`, but we want to convert it
+to an `i32`. Shadowing lets us re-use the `guess` name, rather than forcing us
+to come up with two unique names like `guess_str` and `guess`, or something
+else.
+
+We bind `guess` to an expression that looks like something we wrote earlier:
+
+```rust,ignore
+guess.trim().parse()
+```
+
+Followed by an `ok().expect()` invocation. Here, `guess` refers to the old
+`guess`, the one that was a `String` with our input in it. The `trim()`
+method on `String`s will eliminate any white space at the beginning and end of
+our string. This is important, as we had to press the ‘return’ key to satisfy
+`read_line()`. This means that if we type `5` and hit return, `guess` looks
+like this: `5\n`. The `\n` represents ‘newline’, the enter key. `trim()` gets
+rid of this, leaving our string with just the `5`. The `parse()` method on
```

Comment:
Maybe functions could have a link to their docs (at least the first time they're mentioned).

(Seems like a lot of effort, though.)